include/deal.II/base/memory_space_data.h

```diff
@@ -25,47 +25,83 @@
 #include <functional>
 #include <memory>
 
+#ifdef DEAL_II_USE_KOKKOS_BACKEND
+#  include <Kokkos_Core.hpp>
+#endif
+
 DEAL_II_NAMESPACE_OPEN
 
 /**
  */
 namespace MemorySpace
 {
   /**
-   * Data structure
+   * Structure which stores data on the host or the device depending of the
+   * template parameter @p MemorySpace.
    */
   template <typename Number, typename MemorySpace>
   struct MemorySpaceData
```

Comment:
Can you also document whether the structure *owns* the data it stores (which would include de-allocation when it goes out of scope)?

Separately, I do not see a way to say how much data you actually want to store. The interface only seems to allow copying-into, but does that also allocate memory?

Also separately, I take it that `Number` has to be an arithmetic type, right? Can we add a `static_assert` for this?

Reply:
The primary goal of the structure is to have a single object that stores data on different backend:
 - Host
 - CUDA
 - shared MPI
 - and now Kokkos
 
I have extended the structure to unify the access on the Host and the Kokkos backend. The memory allocation is done where it is needed for each backend. For example, when you are using the CUDA backend you may or may not want to allocate data on the Host too. I could move the allocation if you think that it would be better.